### PR TITLE
Variant management of results

### PR DESCRIPTION
**Pull request by baptsegu**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
New feature : Handle several variants of Crac result extensions

**What is the current behavior?** *(You can also link to an open issue here)*
Several Crac object which can have result extension (i.e Cnec, PstRange, RangeAction, Crac, NetworkAction), containing results for one situation - so called "variant" - for instance situation before RA optimisation.

**What is the new behavior (if this is a feature change)?**
These objects can now have several variants of those results. The variants of the whole Crac are handled by the ResultVariantManager.

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

Method setStep(State state, double step) of the PstRangeResult, which was setting the step and automatically the corresponding tap, has been removed.

### CODE REVIEW

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java [hunk 1]
```diff
@@ -24,24 +26,69 @@
...
+        assertNotNull(pstExtension.getVariant(variantId2));
+
+        // delete one variant
+        variantManager.deleteVariant(variantId2);
 
         assertEquals(1, variantManager.getVariants().size());
+
+        cracExtension = crac.getExtension(ResultExtension.class);
+        cnecExtension = crac.getCnec("Tieline BE FR - Défaut - N-1 NL1-NL3").getExtension(ResultExtension.class);
+        pstExtension = ((PstRange) crac.getRangeAction("PRA_PST_BE")).getExtension(ResultExtension.class);
```
Comment (terriervik) [inline data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java]: What if you delete these lines?

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java [hunk 2]
```diff
@@ -24,24 +26,69 @@
...
+
+        // delete the other variant
+
+        variantManager.deleteVariant(variantId1);
+
+        assertEquals(0, variantManager.getVariants().size());
+
+        cracExtension = crac.getExtension(ResultExtension.class);
+        cnecExtension = crac.getCnec("Tieline BE FR - Défaut - N-1 NL1-NL3").getExtension(ResultExtension.class);
+        pstExtension = ((PstRange) crac.getRangeAction("PRA_PST_BE")).getExtension(ResultExtension.class);
```
Comment (terriervik) [inline data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java]: Same :samoa:

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+@AutoService(ExtensionsHandler.ExtensionSerializer.class)
+public class JsonResultExtension<T extends Identifiable<T>>
+    implements ExtensionsHandler.ExtensionSerializer<T, ResultExtension<T, ? extends Result<T>>> {
+
+    @Override
+    public void serialize(ResultExtension<T, ? extends Result<T>> resultExtension, JsonGenerator jsonGenerator, SerializerProvider serializerProvider) throws IOException {
+        jsonGenerator.writeStartObject();
+
+        // serialize result per result as  the whole map somehow skip the result types
```
Comment (terriervik) [inline data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java]: // serialize result per result as the whole map somehow skips the result types

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @Override
+    public ResultExtension<T, ? extends Result<T>> deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
+
+        ResultExtension<T, ? extends Result<T>> resultExtension = new ResultExtension<>();
+
+        while (!jsonParser.nextToken().isStructEnd()) {
+            switch (jsonParser.getCurrentName()) {
+                case "resultsPerVariant":
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java]: maybe one day

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java [hunk 5]
```diff
@@ -51,33 +52,112 @@ public String getName() {
...
         }
 
-        // todo : in the Result extensions of the Cnec, Crac and RemedialActions, delete the variant with id
-        //        {variantId}
+        if (variants.size() == 1) { // if the crac does not contains other variant than this one : delete all extension
+            getExtendable().removeExtension(ResultExtension.class);
+
+            getExtendable().getCnecs().forEach(cnec -> {
+                cnec.removeExtension(ResultExtension.class);
+            });
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java]: ok

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java [hunk 6]
```diff
@@ -51,33 +52,112 @@ public String getName() {
...
+
+            getExtendable().getNetworkActions().forEach(na -> {
+                na.removeExtension(ResultExtension.class);
+            });
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java]: ok

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java [hunk 7]
```diff
@@ -51,33 +52,112 @@ public String getName() {
...
+                }
+            });
+
+        } else { // else, delete the variants
+
+            getExtendable().getExtension(ResultExtension.class).deleteVariant(variantId);
+
+            getExtendable().getCnecs().forEach(cnec -> {
+                cnec.getExtension(ResultExtension.class).deleteVariant(variantId);
+            });
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java]: ok

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java [hunk 8]
```diff
@@ -51,33 +52,112 @@ public String getName() {
...
+
+            getExtendable().getNetworkActions().forEach(na -> {
+                na.getExtension(ResultExtension.class).deleteVariant(variantId);
+            });
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java]: ok

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/CnecResultTest.java [hunk 9]
```diff
@@ -1,19 +1,9 @@
-/*
- * Copyright (c) 2020, RTE (http://www.rte-france.com)
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/CnecResultTest.java]: :man_facepalming:

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/CnecResultTest.java [hunk 10]
```diff
@@ -1,19 +1,9 @@
...
 package com.farao_community.farao.data.crac_result_extensions;
 
 import org.junit.Test;
 
 import static junit.framework.TestCase.assertEquals;
 
-/**
- * @author Baptiste Seguinot {@literal <baptiste.seguinot at rte-france.com>}
- */
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/CnecResultTest.java]: where am I :male_detective:

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java [hunk 11]
```diff
@@ -51,33 +52,112 @@ public String getName() {
...
+                    pstRa.getExtension(ResultExtension.class).deleteVariant(variantId);
+                }
+            });
+        }
 
         variants.remove(variantId);
     }
 }
+
+
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/ResultVariantManager.java]: you're stronger than the chesktyle

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java (hunk 1 above)
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java]: the test still passes

File: data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java (hunk 2 above)
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/test/java/com.farao_community.farao.data.crac_result_extensions/ResultVariantManagerTest.java]: however these are required as to verify that the extensions were deleted

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java (hunk 3 above)
Comment (baptsegu) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/JsonResultExtension.java]: better

**Comment by baptsegu** [comment]
changes done @terriervik

**Pull request merged**